src/node_eternal_bytes.h

```diff
@@ -0,0 +1,97 @@
+#ifndef SRC_NODE_ETERNAL_BYTES_H_
+#define SRC_NODE_ETERNAL_BYTES_H_
+
+#if defined(NODE_WANT_INTERNALS) && NODE_WANT_INTERNALS
+
+// A pointer to const uint8_t* or const uint16_t* data that can be turned into
+// external v8::String when given an isolate. Unlike UnionBytes, this assumes
+// that the underlying buffer is eternal (will not be garbage collected) and
+// reuses the same v8::String::ExternalStringResourceBase for all materialized
+// strings. This allows the resource to be registered as an external reference
+// for snapshotting.
+
+#include <variant>
+#include "v8.h"
+
+namespace node {
+
+class EternalBytes;
+
+template <typename Char, typename IChar, typename Base>
+class EternalExternalByteResource : public Base {
+  static_assert(sizeof(IChar) == sizeof(Char),
+                "incompatible interface and internal pointers");
+
+ public:
+  explicit EternalExternalByteResource(const Char* data, size_t length)
+      : data_(data), length_(length) {}
+
+  const IChar* data() const override {
+    return reinterpret_cast<const IChar*>(data_);
+  }
+  size_t length() const override { return length_; }
+
+  void Dispose() override {
+    // Do nothing. This class is owned by the EternalBytes instance and so
```

Comment:
If we implement an OwningUnionBytes we don't have to override this - owning version can delete itself and release the strings stored for externalized builtins, while the non-owning UnionBytes with data coming from JS2C can also delete itself because it's just a thin layer that does not actually own the strings.

Reply:
Hm, I'm not sure about this. The overriden implementation of `Dispose` just invokes `delete this`; but here we're allocating the `EternalExternalOneByteResource`/`EternalExternalTwoByteResource` as `static`, so I think we can't just `delete` them?

Reply:
Oh, sorry I missed that the references that need to be registered should be the resources, and thought that the difference is whether they are owned by the bytes class not - the real difference should be instead when their lifetime starts and when their lifetime ends in relation to V8 isolate creation and the string GC.

I think we can just migrate `UnionBytes` to the current structure of `EternalExternalByteResource` and `EternalBytes` - so, the resources reference the data/length/owning_ptr while `UnionBytes` only has a pointer to the resource. The owning external string resources (for externalized builtins) sets the owning_ptr and have their life-cycle tied to a static map (we can probably just turn the `externalized_builtin_sources` map into `std::map<std::string, std::unique_ptr<v8::ExternalStringResourceBase>>` for this), while the non-owning ones are declared static similar to how they are declared in JS2C here. When registering the external references, we just skip the `UnionBytes` for shareble builtins. Then we end up with a simpler `UnionBytes` class, and move the complexity to the different string resources (in this PR we already have that complexity anyway - there are going to be 4 string resource classes no matter what), and we don't need to check two maps in the `BuiltinLoader`.

Reply:
OK, I think I understand. You're proposing something like this (showing only 16-bit strings for simplicity):

```c++
class UnionBytes { Resource *resource_; }

class Resource : v8::String::ExternalStringResource {
  uint16_t* data_;
  size_t length_;
  shared_ptr<void> owning_ptr_;

  void Dispose() override {
    // We ignore Dispose calls from V8, even if we "own" a resource via owning_ptr_.
    // All instantiations of this class are static or owned by a static
    // map, and will be destructed when static variables are destructed.
  }

  // used to ignore externalized builtins for registration
  bool IsDataStatic() { return owning_ptr_ == nullptr; }
}

static std::map<std::string, std::unique_ptr<Resource>> externalized_builtin_sources;

ThreadsafeCopyOnWrite<std::map<std::string, UnionBytes>> sources_;
```

and then we don't need to worry about `resource_` being destructed because it's either initialized statically (for internalized builtins by JS2C) or owned by the static `externalized_builtin_sources` map.

I think that all makes sense to me. It's a little more complicated but certainly less duplicative than this approach. I'll give it a go tomorrow.